


```bash
## Set environmental variables
NEW_DATA_ROOT=./data/downloaded
mkdir -p ${NEW_DATA_ROOT}
CURRENT_DIR=`pwd`
```

```bash
## Download tcga.zip and decompress into $NEW_DATA_ROOT
wget https://hereapp.ccr.cancer.gov/tcga.zip
```


```bash
## Download data from GDC portal 12TB+
GDCEXE=${NEW_DATA_ROOT}/gdc-client  # download gdc-client from TCGA data portal
names=("brca" "paad" "chol" "ucs" "dlbc" "uvm" "ucec" "meso" "acc" "kich" "thym" "tgct" "pcpg" "esca" "sarc" "cesc" "prad" "thca" "ov" "kirc" "blca" "stad" "skcm" "read" "lusc" "luad" "lihc" "lgg" "kirp" "hnsc" "coad" "gbm")
for i in ${!names[@]}; do
  d=${NEW_DATA_ROOT}/tcga_${names[${i}]}/
  cd $d
  for f in "gene_counts" "maf" "dx_svs"; do

    F=$(ls ${f}*.txt)
    # echo $F

    SAVEDIR=${d}all_${f}
    if [ ! -d ${SAVEDIR} ]; then
      mkdir -p $SAVEDIR
      echo "do " $F $SAVEDIR
      $GDCEXE download -m $d/${F} -d $SAVEDIR -n 10
    else
      echo "${SAVEDIR} existed"
    fi
  done
done
cd $CURRENT_DIR
```

```bash
## Decompress MAF files
names=("brca" "paad" "chol" "ucs" "dlbc" "uvm" "ucec" "meso" "acc" "kich" "thym" "tgct" "pcpg" "esca" "sarc" "cesc" "prad" "thca" "ov" "kirc" "blca" "stad" "skcm" "read" "lusc" "luad" "lihc" "lgg" "kirp" "hnsc" "coad" "gbm")
for i in ${!names[@]}; do
  d=${NEW_DATA_ROOT}/tcga_${names[${i}]}/
  cd $d
    for f in `find ${NEW_DATA_ROOT}/tcga_${names[${i}]}/all_maf/ -name "*.gz"`;
    do
    echo $f; gunzip $f;
    done
done
cd $CURRENT_DIR
```

```bash 
##  Check TCGA files
names=("brca" "paad" "chol" "ucs" "dlbc" "uvm" "ucec" "meso" "acc" "kich" "thym" "tgct" "pcpg" "esca" "sarc" "cesc" "prad" "thca" "ov" "kirc" "blca" "stad" "skcm" "read" "lusc" "luad" "lihc" "lgg" "kirp" "hnsc" "coad" "gbm")
for i in ${!names[@]}; do
  PREFIX=$(echo ${names[${i}]} | tr '[:lower:]' '[:upper:]')
  d=${NEW_DATA_ROOT}/tcga_${names[${i}]}
  cd ${d}
  ### check file count
  c1=$(expr $(cat $(find . -maxdepth 1 -name "gene_counts*.txt") | wc -l) - 1)
  c2=$(find ./all_gene_counts/ -name "*_gene_counts.tsv" | wc -l)
  c3=$(expr $(cat $(find . -maxdepth 1 -name "maf*.txt") | wc -l) - 1)
  c4=$(find ./all_maf/ -name "*.gz" | wc -l)
  c41=$(find ./all_maf/ -name "*.maf" | wc -l)
  c5=$(expr $(cat $(find . -maxdepth 1 -name "dx_svs*.txt") | wc -l) - 1)
  c6=$(find ./all_dx_svs/ -name "*.svs" | wc -l)
  if [ $c1 != $c2 ]; then echo "ERROR: check gene_counts files ${c1} ${c2}"; fi
  if [ $c3 == $c4 ] || [ $c3 == $c41 ]; then echo ""; else echo "ERROR: check maf files ${c3} ${c4} ${c41}"; fi
  if [ $c5 != $c6 ]; then echo "ERROR: check svs files ${c5} ${c6}"; fi
done
cd $CURRENT_DIR
```

```bash
## Processing MAF files
for i in ${!names[@]}; do
  bash generate_new_MAFAnnotator.sh ${NEW_DATA_ROOT}/tcga_${names[${i}]} `pwd`
done
cd $CURRENT_DIR
```

```bash
## Generate Merged_FPKM.tsv
for i in ${!names[@]}; do
  PREFIX=$(echo ${names[${i}]} | tr '[:lower:]' '[:upper:]')
  d=${NEW_DATA_ROOT}/tcga_${names[${i}]}
  cd ${d}
  F=$(ls gene_counts_*.txt)
  if [ ! -e gene_counts ]; then
    mkdir gene_counts
    cd gene_counts
    for f in $(find ${d}/all_gene_counts/ -name *.tsv); do
      ln -sf $f .
    done
    cd ..
  fi
  cd $CURRENT_DIR
  python gdc-rnaseq-tool-new.py \
    ${d}/${F} \
    ${d}/gene_counts \
    ${d}
done
cd $CURRENT_DIR
```

```bash
##  Create soft links
names=("brca" "paad" "chol" "ucs" "dlbc" "uvm" "ucec" "meso" "acc" "kich" "thym" "tgct" "pcpg" "esca" "sarc" "cesc" "prad" "thca" "ov" "kirc" "blca" "stad" "skcm" "read" "lusc" "luad" "lihc" "lgg" "kirp" "hnsc" "coad" "gbm")
for i in ${!names[@]}; do
  PREFIX=$(echo ${names[${i}]} | tr '[:lower:]' '[:upper:]')
  d=${NEW_DATA_ROOT}/tcga_${names[${i}]}/
  D=${NEW_DATA_ROOT}/TCGA-$(echo ${names[${i}]} | tr '[:lower:]' '[:upper:]')_256
  echo $D

  if [ ! -d ${D}/svs ]; then
    mkdir -p ${D}/allsvs
    for f in $(find ${d}/all_dx_svs/ -name "*.svs"); do
      ln -sf $f ${D}/allsvs
    done
    ln -sf ${D}/allsvs ${D}/svs
  fi

  if [ $PREFIX == "BRCA" ]; then
    ln -sf ${d}/DataS1.xlsx ${D}/
  fi

  for p in "json" "txt" "csv" "tsv"; do
    for f in $(find ${d} -maxdepth 1 -name "*.${p}"); do
      echo $f
      ln -sf $f ${D}
    done
  done
done
cd $CURRENT_DIR
```

```bash
## Generate Gene Mutation effects files
VERSION="v2"
names=("brca" "paad" "chol" "ucs" "dlbc" "uvm" "ucec" "meso" "acc" "kich" "thym" "tgct" "pcpg" "esca" "sarc" "cesc" "prad" "thca" "ov" "kirc" "blca" "stad" "skcm" "read" "lusc" "luad" "lihc" "lgg" "kirp" "hnsc" "coad" "gbm")
for i in ${!names[@]}; do
  d=${NEW_DATA_ROOT}/tcga_${names[${i}]}/
  if [ -e ${d}/mutation_effects_sorted2${VERSION}.csv ]; then
    rm -rf ${d}/mutation_effects_sorted2${VERSION}.csv
  fi
  if [ -e ${d}/mutation_effects${VERSION}.csv ]; then
    python get_gene_statistic.py ${d} ${VERSION}
  fi
done
cd $CURRENT_DIR
```

```bash
## Generate multi-task labels for training HERE models
# 10 gene mutation classification tasks
# 50 Hallmark gene expression regression tasks
# 5 TIDE-derived gene-related regression tasks
names=("brca" "paad" "chol" "ucs" "dlbc" "uvm" "ucec" "meso" "acc" "kich" "thym" "tgct" "pcpg" "esca" "sarc" "cesc" "prad" "thca" "ov" "kirc" "blca" "stad" "skcm" "read" "lusc" "luad" "lihc" "lgg" "kirp" "hnsc" "coad" "gbm")
MUTATION_VERSION="v2"
PATCH_SIZE=256
for i in ${!names[@]}; do
  PREFIX=$(echo ${names[${i}]} | tr '[:lower:]' '[:upper:]')
  python prepare_dataset_v7.py ${PREFIX} ${MUTATION_VERSION} ${PATCH_SIZE} # add MsigDB hallmark  mean()
  sleep 1
done
cd $CURRENT_DIR
```

```bash
## Build TCGA-ALL2, namely PanCancer
names=("brca" "paad" "chol" "ucs" "dlbc" "uvm" "ucec" "meso" "acc" "kich" "thym" "tgct" "pcpg" "esca" "sarc" "cesc" "prad" "thca" "ov" "kirc" "blca" "stad" "skcm" "read" "lusc" "luad" "lihc" "lgg" "kirp" "hnsc" "coad" "gbm")
mkdir -p ${NEW_DATA_ROOT}/TCGA-ALL2_256/allsvs/
for i in ${!names[@]}; do
  PREFIX=$(echo ${names[${i}]} | tr '[:lower:]' '[:upper:]')
  D=${NEW_DATA_ROOT}/TCGA-${PREFIX}_256/allsvs
  for f in $(ls ${D}/*.svs); do
    ln -sf $f ${NEW_DATA_ROOT}/TCGA-ALL2_256/allsvs/
  done
  echo $i $PREFIX
done
```

```bash
## Create patches, download CLAM from https://github.com/mahmoodlab/CLAM
cd ./CLAM
for i in ${!names[@]}; do
  PREFIX=$(echo ${names[${i}]} | tr '[:lower:]' '[:upper:]')

  PROJ_NAME=TCGA-${PREFIX} 
  DATA_DIRECTORY=${NEW_DATA_ROOT}/${PROJ_NAME}_${PATCH_SIZE}/svs
  RESULTS_DIRECTORY=${NEW_DATA_ROOT}/${PROJ_NAME}_${PATCH_SIZE}/
  python create_patches_fp.py \
  --source ${DATA_DIRECTORY} \
  --save_dir ${RESULTS_DIRECTORY} \
  --patch_size ${PATCH_SIZE} \
  --preset tcga.csv \
  --seg --patch --stitch

done
cd $CURRENT_DIR
```

```bash
## Download backbone models into ./backbones directory
# PLIP, ProvGigaPath, CONCH, UNI
Please see [`backbones/README.md`](./backbones/README.md) for details.
```

















